### PR TITLE
Feature/unit testing

### PR DESCRIPTION
**Pull request by MacPatric**
This PR introduces unit testing to the project using `Vitest`. 

I used the code already existing in the MetropolCon 2026 version and removed all references to translations with react-i18n (which will be implemented later).

Additionally I made three minor improvements.

### Minor Improvements
- Fixed `class` → `className` in `Switch.jsx` for consistency with jsx conventions
- Updated `People.jsx` to use the configurable search label from `config.json`
- Added IDE settings (`.idea/`) to `.gitignore`

### Unit Testing
- Added `Vitest` as the test runner with jsdom environment
- Added testing libraries: `@testing-library/react` and `@testing-library/jest-dom` for component testing
- Created `src/setupTests.js` for global test setup
- Updated `vite.config.js` with test configuration
- Added npm scripts: test (start the tests with `npm test`)
- Added tests for many components, but not all. 

The tests are all in separate files, each file follows the naming convention `<component name>.test.js`. Dependencies are mocked to ensure the tests run in isolation without external state.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -23,7 +23,10 @@
   "scripts": {
     "start": "vite",
     "build": "vite build",
-    "preview": "vite preview"
+    "preview": "vite preview",
+    "test": "vitest",
+    "test:watch": "vitest watch",
+    "test:coverage": "vitest run --coverage"
```
Comment (aJanuary) [inline package.json:30]: [❓Question](https://evertz.atlassian.net/wiki/spaces/~701219ae3e524169d4e55afd643b7d8c54fdb/pages/733808663/Code+review+comment+types#%E2%9D%93-Question): There is no coverage dependency declared, so when I run this I get  `MISSING DEPENDENCY  Cannot find dependency '@vitest/coverage-v8'`. It _does_ prompt me if I want to install it, so I wanted to ask if this was a deliberate choice to make installing the coverage dependency lazy, or if it was an oversight and we should add a dev dependency on `@vitest/coverage-v8`.

File: vite.config.js [hunk 2]
```diff
@@ -104,4 +104,10 @@ export default defineConfig({
     // the warning limit.
     chunkSizeWarningLimit: 1000,
   },
+  test: {
```
Comment (aJanuary) [inline vite.config.js:107]: [Suggestion](https://evertz.atlassian.net/wiki/spaces/~701219ae3e524169d4e55afd643b7d8c54fdb/pages/733808663/Code+review+comment+types#Suggestion): Add `clearMocks: true`.

Because the mocks are declared in the describe scope and re-used between tests, they are retaining state between tests. This means that if, in the future, we ever try to asset on specifically how many times a method has been called rather than just whether or not it has been called, the number of calls will depend on what tests have run before it.

Adding `clearMocks: true` will reset the mocks between each test, making them isolated from each other. From my local testing, it has no measurable impact on the time it takes to run the tests.

File: src/components/ProgramItem.test.jsx [hunk 3]
```diff
@@ -0,0 +1,522 @@
...
+vi.mock('react-icons/io5', () => ({
+  IoChevronDownCircle: () => <span data-testid="chevron-icon">Chevron</span>,
+}));
+
+vi.mock('react-icons/hi', () => ({
+  HiLink: () => <span data-testid="link-icon">Link</span>,
+}));
+
+// Suppress console.log for tests
+global.console.log = vi.fn();
```
Comment (aJanuary) [inline src/components/ProgramItem.test.jsx]: [❗Important](https://evertz.atlassian.net/wiki/spaces/~701219ae3e524169d4e55afd643b7d8c54fdb/pages/733808663/Code+review+comment+types#%E2%9D%97-Important): This changes the global console.log, but never restores it. This means if we ever genuinely want to use the console.log output as a diagnostic tool in the tests, we're going to tear our hair out trying to figure out why we can't see anything.

I think we can just remove this now. Nothing in ProgramItem logs to the console anymore.

**Review by aJanuary — CHANGES_REQUESTED** [review]
Amazing work! Will be great to have a test suite to give us some more confidence.

I do have a few concerns about the implementation. Aside from the one about mocking console.log, I think we can reasonably proceed with the PR as-is and schedule the rest for follow-up.

### Mocking child components
My preference would be against mocking child components in most cases.

Mocking a child component only proves you passed _some_ value, but it doesn't catch typos in prop names or where the child expects a different shape than the parent provides it. Rendering with a real child exercises the actual contract between them. If a child's contract changes, the mocked version is going to keep returning the old fake output and the parent's test will stay green, even though the app is broken.

It also reduces the amount of maintenance we have to do on the mocks to keep them in sync.

By mocking the child components, we're testing the implementation details - was this sub-component called with this prop. When really we want to be testing user visible behaviour.

I think we would benefit from having the tests more integration-flavoured than hardline isolated tests on the components.

As a blatant appeal to authority, Kent C. Dodds - the creator of the testing library we're using - is in the "integration over mocks" camp https://www.software-engineering-unlocked.com/no-mocks-allowed/

### Config duplication
You're mocking the config from scratch in each test. I'm not actually against that per-se. I think people have a tendency to jump too quickly to extracting out any superficial similarities in test code.

But you're also only defining a minimal config needed for that specific test suite, ignoring config keys that are required by the application as a whole. I think this leads to brittle tests. If a config key is already required, then new code in existing components should be able to use it fearlessly. You don't want unrelated tests to break because the code didn't mock the config value because the old code didn't need it. As a general philosophy, I prefer to create a realistically complete environment for the tests to run, rather than hyper-focussing on what the current code happens to need.

And if we are defining an entire config file for each component, we then start to loose the ability to tell what pieces of this config are important for this test and what are there to create the realistic environment.

To that end, the pattern I currently favour is creating a helper to construct the mocks and fakes that configures it with a bunch of reasonable defaults, and then overriding behaviour needed for the specific tests.

File: package.json (hunk 1 above)
Comment (MacPatric) [reply package.json:30]: I actually did overlook this and have now added the dependency to `package.json`.

File: src/components/ProgramItem.test.jsx (hunk 3 above)
Comment (MacPatric) [reply src/components/ProgramItem.test.jsx]: This was unintentional. As a long-time software developer, I’ve accumulated a lot of reusable code snippets over the years. In this case, it was simply a copy-paste error, which I’ve now corrected. The line has been removed.

File: vite.config.js (hunk 2 above)
Comment (MacPatric) [reply vite.config.js:107]: I’ve added `clearMocks: true` to the test configuration.

**Comment by MacPatric** [comment]
### Mocking Child Components and Config Duplication

Both topics are part of the same discussion: *what* and *how* we test. Should we write unit tests or integration tests? Kent Dodds is absolutely right when it comes to integration testing. However, he’s wrong—at least in my view—when it comes to test-driven development and unit testing. After 30 years in the industry, I feel confident saying that. I’ve earned a bit of authority on this myself. ;-)

As a test-driven developer, I often (admittedly not always) start by writing unit tests and then implement the business logic until my tests pass. When I write unit tests, I’m deliberately *not* testing user-visible behavior—I’m testing implementation details. In my experience, precise and error-free implementation is more critical to a system’s stability than the part of the software the user ultimately sees.

I personally write very few integration tests, and only at the very end of development, when I truly feel the need to test the interaction between my components, services, classes, or functions. Often, this isn’t necessary if the unit tests cover all cases, including edge cases. Admittedly, that’s not the case here: for the MetropolCon 2026, I focused on the components and functions I *had* to change, and I had to ignore the rest due to time constraints.

I use mocks—especially for `config.json`—so that when I read the tests, I can immediately see what data is actually being tested. I don’t need a complete `config.json` for each test; I only need the relevant portion. That’s why my test classes often include a mocked `config.json`, but only with the data actually required for the tests. This approach has proven effective in my professional practice for many years, even if it occasionally violates the DRY principle in the test classes.

That’s the software-philosophical side of the discussion. However, the decision on how to proceed ultimately lies with you and James as the project maintainers. If you say you want fewer unit tests and more integration tests, that’s the direction we’ll take.

The question is: how do we handle the existing unit tests? They aren’t integration tests, but they *are* functional unit tests that cover some important implementation details. James and I have agreed that my next steps will include preparing a PR for my changes regarding react-i18n and multilingual support. I’d like to use the current tests for this, as they already exist in nearly identical form in [*conclar-metropolcon-2026*](https://github.com/MacPatric/conclar-metropolcon-2026)—including multilingual support. In the future, once the MetropolCon changes are integrated into ConClar, we can revisit and refine the tests.

What are your thoughts on the next steps?

**Comment by lostcarpark** [comment]
Thank you so much for your work on this. I've had an initial look and run the tests locally, and it looks like a great start.

My partner, Fionna, works as an Automation Architect. We're both new to ViTest, but we'll review over the weekend.

I am inclined to favour merging the tests unless we find any major issues with them. We can try to increase test coverage and refine them after that.

File: src/ProgramData.test.js [hunk 4]
```diff
@@ -0,0 +1,528 @@
...
+      ];
+
+      const people = [
+        { id: 'p1', name: 'Alice', sortname: 'Alice' },
+        { id: 'p2', name: 'Zoe', sortname: 'Zoe' },
+      ];
+
+      ProgramData.addProgramParticipantDetails(program, people);
+
+      expect(program[0].people[0].name).toBe('Alice');
```
Comment (lostcarpark) [inline src/ProgramData.test.js]: Because the test sets up the name and sortname to be the same, this doesn't in fact test that they are sorted by sortname - the test would still pass if the participants were sorted by name

**Review by lostcarpark — CHANGES_REQUESTED** [review]
Fionna and I have reviewed the tests, and we think  on the whole they are a great starting point.

We've just highlighted one that we think could use a tweak. The `sortname` and `name` values should be different so they will sort differently (we also think adding at least one more item to the people array would be useful).

We'd love to see more test coverage added, but with the change above, we can get this merged and then work on adding additional cover.

**Comment by lostcarpark** [comment]
I think unit tests are very useful for testing isolated functionality, but once we get that in place, an additional set of tests that verify interaction of different components would be very useful.
I'd also love to see some functional testing using something like PlayWright.

File: src/ProgramData.test.js (hunk 4 above)
Comment (MacPatric) [reply src/ProgramData.test.js]: You were right, the test would passed when sorted by name. I changed the data and the assertions to make the test explicitly check for the expected sortnames at indices 0 and 1.

**Review by lostcarpark — APPROVED** [review]
Looks good now. Ready to merge.

**Comment by lostcarpark** [comment]
Oops not quite ready - fixed conflict in dependencies.

File: vite.config.js [hunk 5]
```diff
@@ -104,4 +104,11 @@ export default defineConfig({
...
+    globals: true,
+    environment: 'jsdom',
+    setupFiles: ['./src/setupTests.js'],
+    css: true,
+    cleanMocks: true
```
Comment (lostcarpark) [inline vite.config.js:112]: I don't think this is a valid ViteTest option... should it be clearMocks?
https://vitest.dev/config/clearmocks

File: src/components/Footer.test.jsx [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+  it('renders site note markdown content', () => {
+    render(<Footer />);
+
+    expect(screen.getByText('Test site note with **markdown**')).toBeDefined();
+  });
+
+  it('renders footer bottom section', () => {
+    const { container } = render(<Footer />);
+
+    const footerBottom = container.querySelector('.footer-bottom');
```
Comment (lostcarpark) [inline src/components/Footer.test.jsx:50]: `.footer-bottom` is no longer exists in the footer area (just `.footer-site`, `.footer-copyright`, and `.footer-conclar`), so this returns NULL, but it seems to pass anyway. I think this questions the validity of this pattern for checking a section exists, so I think this needs to be checked.

File: src/utils/JsonParse.test.js [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+    });
+
+    it('should ignore text before start position', () => {
+      const input = 'skip this {"key": "value"}';
+      const result = JsonParse.extractJson(input, 10);
+      expect(result).toEqual([{ key: 'value' }]);
+    });
+
+    it('should handle JSON with line comments', () => {
+      const input = '{/* comment */ "key": "value"}';
```
Comment (lostcarpark) [inline src/utils/JsonParse.test.js:44]: Test title says "line comments" but `/* ... */` is a block comment.

File: src/utils/JsonParse.test.js [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+      const result = JsonParse.extractJson(input);
+      expect(result).toEqual([{ key: 'value' }]);
+    });
+
+    it('should handle JSON with block comments', () => {
```
Comment (lostcarpark) [inline src/utils/JsonParse.test.js:49]: Likewise, test says `block comments`, but `//` is a line comment. Just need to swap the test title with the test above.

File: src/components/Navigation.test.jsx [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+
+  it('uses NavLink for internal routes', () => {
+    const { container } = render(
+      <BrowserRouter>
+        <Navigation />
+      </BrowserRouter>
+    );
+
+    const programLink = screen.getByText('Programme');
+    expect(programLink.className).toContain('');
```
Comment (lostcarpark) [inline src/components/Navigation.test.jsx:156]: Every string contains the empty string, so this will always pass. I think the intention is to check `programLink` has no class. How about `expect(programLink.classList.length).toBe(0);`?

File: src/components/HelpText.test.jsx [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+      CLOSE_ARIA_LABEL: 'Dismiss help text',
+    },
+  },
+}));
+
+// Mock easy-peasy
+vi.mock('easy-peasy', () => ({
+  useStoreState: vi.fn(),
+  useStoreActions: vi.fn(),
+}));
```
Comment (lostcarpark) [inline src/components/HelpText.test.jsx:20]: This is repeated across several components. Even though it's small, it would be better to move into a mock class.

File: src/components/Settings.test.jsx [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+
+    const radioLight = screen.getByLabelText('settings.dark_mode.light_mode_label');
+    fireEvent.click(radioLight);
+    expect(mockSetDarkMode).toHaveBeenCalledWith('light');
+  });
+
+  it('handles dark mode change to browser default', () => {
+    // Override state: start with dark mode set to 'light' so we can change to 'browser'
+    useStoreState.mockImplementation((selector) => {
+      const state = {
```
Comment (lostcarpark) [inline src/components/Settings.test.jsx:212]: I think this should mock other properties than just darkMode, like the earlier mock above, as this will cause unknown values to be passed to the settings form.

**Review by lostcarpark — CHANGES_REQUESTED** [review]
I've found some issues that I think need investigation, and possible rework.

Could you take a look at the sections I've highlighted?

Thanks!

File: src/components/Footer.test.jsx (hunk 6 above)
Comment (lostcarpark) [reply src/components/Footer.test.jsx:50]: I think a better pattern to check element is actually present is `expect(footerBottom).toBeInTheDocument();`.

File: vite.config.js (hunk 5 above)
Comment (MacPatric) [reply vite.config.js]: Of course it should be clearMocks. Sorry about the typo.

File: src/components/Footer.test.jsx (hunk 6 above)
Comment (MacPatric) [reply src/components/Footer.test.jsx]: I learned something here: `toBeDefined()` passes for null, it only fails on undefined. I replaced `toBeDefined()`with `toBeInTheDocument()` in every test that checks for DOM elements.

File: src/utils/JsonParse.test.js (hunk 8 above)
Comment (MacPatric) [reply src/utils/JsonParse.test.js]: I swapped the test titles.

File: src/components/Navigation.test.jsx (hunk 9 above)
Comment (MacPatric) [reply src/components/Navigation.test.jsx]: `classList.length().toBe(0)` would not work in this case. NavLink is active and the classList length would be 1. As this test checks if NavLink is active I use `expect(programLink).toHaveClass('active')` now as only NavLink's `isActive` callback can produce this.

File: src/components/HelpText.test.jsx (hunk 10 above)
Comment (MacPatric) [reply src/components/HelpText.test.jsx]: You are right, I violated the DRY principle here, which could be justified in some test cases. But in this case it is unnecessary and I moved the mock to `setupTests.js`.

File: src/components/Settings.test.jsx (hunk 11 above)
Comment (MacPatric) [reply src/components/Settings.test.jsx:206]: The complete state is mocked now to avoid undefined or unknown values in the rendered form.
